waiter/src/waiter/auth/saml.clj

```diff
@@ -0,0 +1,195 @@
+                            (catch Exception e
+                              (throw (ex-info "Could not parse saml-auth-data." {:status 400
+                                                                                 :saml-auth-data saml-auth-data
+                                                                                 :inner-exception e}))))
+                          t-now (t/now)
+                          _ (when-not (t/before? t-now not-on-or-after)
+                              (throw (ex-info "Could not authenticate user. Expired SAML assertion."
+                                              {:status 400
+                                               :saml-assertion-not-on-or-after not-on-or-after
+                                               :t-now t-now})))
+                          age-in-seconds (t/in-seconds (t/interval t-now (t/min-date (t/plus t-now (t/days 1)) not-on-or-after)))
+                          {:keys [authorization/principal authorization/user] :as auth-params-map}
+                          (auth/auth-params-map saml-principal)
+                          request' (-> request
+                                       (assoc :request-method :get :body nil :content-type nil :content-length nil)
+                                       (update :headers dissoc "content-length" "content-type"))]
+                      (auth/handle-request-auth request-handler request' principal auth-params-map password age-in-seconds))
+                    (throw (ex-info "Invalid request method for use with SAML authentication"
+                                    {:log-level :info :request-method request-method :status 405})))
+            (throw (ex-info "Invalid request method for use with SAML authentication"
+                            {:log-level :info :request-method request-method :status 405}))))))))
+
+(defn certificate-x509
+  "Takes in a raw X.509 certificate string, parses it, and creates a Java certificate."
+  [x509-string]
+  (let [fty (java.security.cert.CertificateFactory/getInstance "X.509")
+        bais (io/input-stream (.getBytes x509-string))]
+    (.generateCertificate fty bais)))
+
+(defn validate-saml-assertion-signature
+  "Checks that the SAML assertion has a valid signature."
+  [assertion idp-cert]
+  (if-let [signature (.getSignature assertion)]
+    (let [idp-pubkey (-> idp-cert certificate-x509 saml-shared/jcert->public-key)
+          public-creds (doto (new org.opensaml.xml.security.x509.BasicX509Credential)
+                         (.setPublicKey idp-pubkey))
+          validator (new org.opensaml.xml.signature.SignatureValidator public-creds)]
+      (try
+        (.validate validator signature)
+        true
+        (catch org.opensaml.xml.validation.ValidationException ex
+          (log/warn "Signature NOT valid" (.getMessage ex))
+          false)))
+    false ;; if not signature is present
```

Comment:
Do you need to explicitly return `false` from this method?

Reply:
done